Wähle zwei Primzahlen, möglichst gross. In unserem Beispiel wählen wir $p=47$ und $q=59$. Daraus bilden wir das Produkt $n=p*q$

In [1]:
p=47
q=59
n=p*q
print('Das Produkt lautet n =', n)

Das Produkt lautet n = 2773


Berechne nun die Euler'sch $\phi$-Funktion: $\phi=(p-1)(q-1)$

In [2]:
phi=(p-1)*(q-1)
print('Der Eulersche Phi-Funktionswert lautet phi = ', phi)

Der Eulersche Phi-Funktionswert lautet phi =  2668


Wähle nun eine Zahl $e$ (Primzahl), die zu $\phi(n)$ teilerfremd ist, also weder ein Teiler von $46$ noch von $58$ ist. Wir wählen $e=17$.
Dann löse die Gleichung:

$$d\cdot e=mod \phi$$

Das ist gleichbedeutend mit

$$d=\frac{1+(q-1)\cdot (p-1)}{e}$$



In [3]:
e=17 # scheint nur mit 17 zu funktionieren!?
d=(1+(q-1)*(p-1))/e
d=int(d) # Umwandlung in Integer notwendig, sonst gibt es Rundungsfehler
print('Die Lösung von d*e mod phi = 1 ist: d= ', d)

Die Lösung von d*e mod phi = 1 ist: d=  157


Damit haben wir einen öffentlichen Schlüssel $(e, n)=(17, 2773)$ und einen privaten Schlüssel $(d, n)=(157, 2773)$ erzeugt.

Nun verschlüsseln wir mit dem privaten Schlüssel eine Botschaft (Message) $m$, z. B. für $m=1220$.

Dazu berechnen wir:

$$ x=m^e mod \quad n$$

In [4]:
m=1220
x=m**e%n
print('Die verschlüsselte Zahl ist x = ', x)

Die verschlüsselte Zahl ist x =  1750


Zur Entschlüsselung berechnen wir mit dem geheimen Schlüssel $(d, n)=(157, 2773)$:

$$mr=x^d mod\quad n$$

In [5]:
mr=x**d%n
print('Die rekonstruierte Zahl ist: mr = ', mr)

Die rekonstruierte Zahl ist: mr =  1220


In [37]:
print(e, d)

17 157
